src/remote/fs/fs.go

```diff
@@ -0,0 +1,210 @@
+// Package fs provides an io/fs.FS implementation over the remote API
+package fs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	iofs "io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/client"
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/digest"
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"google.golang.org/protobuf/proto"
+)
+
+type Client interface {
+	ReadBlob(ctx context.Context, d digest.Digest) ([]byte, *client.MovedBytesMetadata, error)
+}
+
+// fs is an io/fs.FS implemented on top of a REAPI directory. This will download files on demand, as they are needed.
+type fs struct {
+	c    Client
+	root *pb.Directory
+}
+
+// New creates a new filesystem on top of the given proto, using client to download files on demand.
+func New(c Client, root *pb.Directory) iofs.FS {
+	return &fs{
+		c:    c,
+		root: root,
+	}
+}
+
+// Open opens the file with the given name
+func (fs *fs) Open(name string) (iofs.File, error) {
+	return fs.open(".", filepath.Clean(name), fs.root)
+}
+
+func (fs *fs) open(path, name string, wd *pb.Directory) (iofs.File, error) {
+	name, rest, hasToBeDir := strings.Cut(name, string(filepath.Separator))
+	// Must be a dodgy symlink that goes past our root.
+	if name == ".." || name == "." {
+		return nil, os.ErrNotExist
+	}
+
+	for _, d := range wd.Directories {
```

Comment:
You probably want to use `GetTree` instead of walking every directory one level at a time? It will be a bit slower up front but should overall be faster by using one RPC to fetch the whole tree instead of many.

Reply:
I think a lot of the time, I'm going to be downloading just one file in the whole directory tree (i.e. the BUILD file we're parsing). How does `GetTree` work? Doesn't that download the whole thing in one go?

Reply:
Oh, GetTree is done server side? That should be faster then!